usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripServiceId.java

```diff
@@ -0,0 +1,74 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.usecase;
+
+import org.apache.logging.log4j.Logger;
+import org.mobilitydata.gtfsvalidator.domain.entity.gtfs.Calendar;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.error.ServiceIdNotFoundNotice;
+import org.mobilitydata.gtfsvalidator.usecase.port.GtfsDataRepository;
+import org.mobilitydata.gtfsvalidator.usecase.port.ValidationResultRepository;
+
+import java.util.Collection;
+import java.util.Set;
+
+/**
+ * Use case for E036 to validate that all records of `trips.txt` refer to an existing {@code Calendar} or
+ * {@code CalendarDate} from files `calendar.txt` or `calendar_dates.txt
+ */
+public class ValidateTripServiceId {
+    private final GtfsDataRepository dataRepo;
+    private final ValidationResultRepository resultRepo;
+    private final Logger logger;
+
+    /**
+     * @param dataRepo   a repository storing the data of a GTFS dataset
+     * @param resultRepo a repository storing information about the validation process
+     * @param logger     a logger to log information about the validation process
+     */
+    public ValidateTripServiceId(final GtfsDataRepository dataRepo,
+                                 final ValidationResultRepository resultRepo,
+                                 final Logger logger) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+        this.logger = logger;
+    }
+
+    /**
+     * Use case execution method: Checks if `service_id` of a trip refers to a record from files `calendar.txt` or
+     * `calendar_dates.txt`. A new notice is generated each time this condition is false.
+     * This notice is then added to the {@link ValidationResultRepository} provided in the constructor.
+     */
+    public void execute() {
+        logger.info("Validating rule E036 - `service_id` not found" + System.lineSeparator());
+
+        final Collection<Calendar> calendarCollection = dataRepo.getCalendarAll();
+        final Set<String> calendarDatesKeyCollection = dataRepo.getCalendarDateAll().keySet();
+        dataRepo.getTripAll().forEach(trip -> {
+            final String serviceId = trip.getServiceId();
+            if (calendarCollection.stream().noneMatch(calendar -> calendar.getServiceId().equals(serviceId)) &&
+                    calendarDatesKeyCollection.stream()
+                            .noneMatch(calendarDateKey -> calendarDateKey.startsWith(serviceId))) {
```

Comment:
Shouldn't this be `.equals` instead of `.startsWith`?

Reply:
For `CalendarDate`,  entities are mapped on a key made of the concatenation of fields `service_id` and `date`, so it makes sense to check that the key starts with a given `service_id`. A solution could be to store `CalendarDate` entities as a nested map where the first key would be `service_id` and the second `date`, I think that this would help keeping consistency with the code base.

Reply:
As discussed, having maps would be less memory efficient for a marginal gain in that case.
We will
- add a comment/javadoc explicitely stating the expected form of keys
- add a test when key are generated (entity add to the repo) to check they are of expected form

Reply:
Ah, yes, thanks for the reminder. Yes, in retrospect, I agree and think that's a better design, primarily because you gain a lot of efficiency here (which I was also about to comment on above).

Right now you're lot levering the benefits of HashMaps at all - both calendar and calendar_dates are coverted to collection/set and you're iterating though them looking for matches.  When trips.txt and calendar_dates.txt are both huge, this is very inefficient.

If you get both `calendars` and `calendar_dates` as HashMaps here keyed on `service_id`, then it becomes a simple constant `.contains()` call to check if the service_id exists:

`if (!calendars.contains(service_id) && !calendar_dates.contains(service_id) {`

We were originally trying to optimize the case of checking for duplicates in calendar_dates.txt by creating the composite key, but given the above, I think it's worth changing. This is likely true for other composite keys when you need to reference the record from another table based on part of the composite key, although I'd wait to change any implementations until we run across the need to.

Reply:
As discussed, because in the majority of feeds have a relatively big trips file compared to calendar_date, the memory/performance tradeoff leans towards implementing @lionel-nj suggestion and have a constant lookup time.